lib/Sema/TypeCheckType.cpp

```diff
@@ -1798,6 +1798,18 @@ static Type applyNonEscapingIfNecessary(Type ty,
   return ty;
 }
 
+static Type applySendableIfNecessary(Type ty, TypeResolutionOptions options) {
+  if (!options.is(TypeResolverContext::FunctionInput)) {
+    return ty;
+  }
+
+  auto *funcTy = ty->castTo<FunctionType>();
+  auto extInfo = funcTy->getExtInfo();
+  assert(!extInfo.isSendable());
```

Comment:
What happens if you apply `@Sendable` to a typealias whose underlying function type is already Sendable? I don't think we should crash in an asserts build. As of today for the non-typealias case we give an error:

```
error: repl.swift:1:19: error: duplicate attribute
let f : @Sendable @Sendable () -> Void = {@Sendable in ()}
```

perhaps we need something similar here? I could be convinced to go without an error diagnostic, since it could be an innocuous mistake, but pointing it out to the developer may help them realize their assumptions are wrong. Thoughts?

Relatedly, what happens if the typealias is marked `@preconcurrency` (e.g., a `@preconcurrency` import of a module)? Would users later get a compiler crash or error diagnostic if the library author who vended that typelias updated it to adopt concurrency? For example, they added `@Sendable` onto the function type. If we decided to make the duplicated attribute on it an error, that adoption causing clients to break would probably be bad.

Reply:
Warning until Swift 6 seems reasonable. But I kind of think this should be a language proposal. I wonder if people would also want this for global actors (i.e., `@MainActor`) or other kinds of type attributes. Limiting it to just Sendable feels like a narrow carve-out that might not be intuitive.

Reply:
Hmm okay, I will close this PR for now and will start a thread on the forums to discuss it further.